plugins/outputs/dynatrace/dynatrace.go

```diff
@@ -0,0 +1,255 @@
+		return "", fmt.Errorf("error normalizing the string: %s", s)
+	}
+	return normalizedString, nil
+}
+
+func (d *Dynatrace) escape(v string) string {
+	return strconv.Quote(v)
+}
+
+func (d *Dynatrace) Write(metrics []telegraf.Metric) error {
+	var buf bytes.Buffer
+	var tagb bytes.Buffer
+	if len(metrics) == 0 {
+		return nil
+	}
+
+	for _, metric := range metrics {
+		// first write the tags into a buffer
+		tagb.Reset()
+		if len(metric.Tags()) > 0 {
+			for tk, tv := range metric.Tags() {
+				tagKey, err := d.normalize(tk, maxDimKeyLen)
+				if err != nil {
+					continue
+				}
+				fmt.Fprintf(&tagb, ",%s=%s", strings.ToLower(tagKey), d.escape(tv))
+
+			}
+		}
+		if len(metric.Fields()) > 0 {
+			for k, v := range metric.Fields() {
+				var value string
+				switch v := v.(type) {
+				case string:
+					continue
+				case float64:
+					if !math.IsNaN(v) && !math.IsInf(v, 0) {
+						value = fmt.Sprintf("%f", v)
+					} else {
+						continue
+					}
+				case uint64:
+					value = strconv.FormatUint(v, 10)
+				case int64:
+					value = strconv.FormatInt(v, 10)
+				case bool:
+					if v {
+						value = "1"
+					} else {
+						value = "0"
+					}
+				default:
+					d.Log.Infof("Dynatrace type not supported! %s", v)
+					continue
+				}
+
+				// metric name
+				metricKey, err := d.normalize(k, maxMetricKeyLen)
+				if err != nil {
+					continue
+				}
+
+				metricID, err := d.normalize(d.Prefix+metric.Name()+"."+metricKey, maxMetricKeyLen)
+				// write metric name combined with its field
+				if err != nil {
+					continue
+				}
+				fmt.Fprintf(&buf, "%s", metricID)
+				// add the tag string
+				if len(tagb.String()) > 0 {
+					fmt.Fprintf(&buf, "%s", tagb.String())
+				}
+
+				// write measured value
+				fmt.Fprintf(&buf, " %v\n", value)
+			}
+		}
+	}
+
+	return d.send(buf.Bytes())
+}
+
+func (d *Dynatrace) send(msg []byte) error {
+	var err error
+	req, err := http.NewRequest("POST", d.EnvironmentURL, bytes.NewBuffer(msg))
+	if err != nil {
+		d.Log.Errorf("Dynatrace error: %s", err.Error())
+		return fmt.Errorf("Dynatrace error while creating HTTP request:, %s", err.Error())
+	}
+	req.Header.Add("Content-Type", "text/plain; charset=UTF-8")
+
+	if len(d.EnvironmentAPIToken) != 0 {
+		req.Header.Add("Authorization", "Api-Token "+d.EnvironmentAPIToken)
+	}
+	// add user-agent header to identify metric source
+	req.Header.Add("User-Agent", "telegraf")
+
+	resp, err := d.client.Do(req)
+	if err != nil {
+		d.Log.Errorf("Dynatrace error: %s", err.Error())
+		fmt.Println(req)
+		return fmt.Errorf("Dynatrace error while sending HTTP request:, %s", err.Error())
+	}
+	defer resp.Body.Close()
+
+	// print metric line results as info log
+	if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusAccepted {
+		bodyBytes, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			d.Log.Errorf("Dynatrace error reading response")
+		}
+		bodyString := string(bodyBytes)
+		d.Log.Infof("Dynatrace returned: %s", bodyString)
```

Comment:
Logger is too noisy. When working properly, almost nothing should be written to `Log.Info`. Probably suitable as a `Log.Debug`

Reply:
Changed